### PR TITLE
Migrate thnn_conv2d from THC to ATen

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63505
* #63504
* #63429
* __->__ #63428

Closes gh-24644, closes gh-24645

Differential Revision: [D30441307](https://our.internmc.facebook.com/intern/diff/D30441307)

### CODE REVIEW

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu [hunk 1]
```diff
@@ -1,12 +1,525 @@
...
+  int dimw = 2;
+
+  if (ndim == 4) {
+    dimf++;
+    dimh++;
+    dimw++;
+  }
+
+  // Allow for empty batch size but not other dimensions
+  const bool valid_empty = c10::multiply_integers(in_sizes.slice(1)) != 0;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ConvolutionMM2d.cu:46]: This is not right if ndim is 3 (the you need in_sizes.numel() > 0)? Btw, do we still allow 3d inputs, or is this code path never exercised?

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu [hunk 2]
```diff
@@ -1,12 +1,525 @@
...
+              "Given input size per channel: ",
+              IntArrayRef{inputHeight, inputWidth},
+              ". Calculated output size per channel: ",
+              IntArrayRef{outputHeight, outputWidth},
+              ". Output size is too small");
+
+  if (weight.defined()) {
+    const auto w_sizes = weight.sizes();
+    int64_t nInputPlane = w_sizes[1];
+    if (w_sizes.size() == 2) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ConvolutionMM2d.cu:76]: can this happen? I though weight shape is checked to have 4 dims before dispatching to slow conv?

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu [hunk 3]
```diff
@@ -1,12 +1,525 @@
...
+
+      // Do Bias first:
+      // M,N,K are dims of matrix A and B
+      // (see http://docs.nvidia.com/cuda/cublas/#cublas-lt-t-gt-gemm)
+      int64_t m_ = nOutputPlane;
+      int64_t n_ = outputHeight * outputWidth;
+      int64_t k_ = 1;
+
+      // Do GEMM (note: this is a bit confusing because gemm assumes column-major matrices)
+      if (bias.defined()) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ConvolutionMM2d.cu:205]: is it just expanding bias to output size? Should we be able to do the same with `.expand()` and `output_n.copy_`, without gemm call and `ones`?

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu [hunk 4]
```diff
@@ -1,12 +1,525 @@
...
+            scalar_t(1),
+            gemm_in_ptr, k,
+            grad_output_n.data_ptr<scalar_t>(), k,
+            scalar_t(1),
+            grad_weight.data_ptr<scalar_t>(), n
+        );
+      }
+
+      // Do Bias:
+      if (grad_bias.defined()) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ConvolutionMM2d.cu:443]: This also should be doable by just a `sum` call with appropriate dimensions?

**Review by ngimel — COMMENTED** [review]
This looks great @peterbell10, I left a few comments.

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/ConvolutionMM2d.cu:46]: You're right, `_convolution` enforce the shape check before this is ever called. Should I just remove all the `dim=3` handling code?

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu (hunk 3 above)
Comment (peterbell10) [reply aten/src/ATen/native/cuda/ConvolutionMM2d.cu:205]: Is it alright if I leave algorithmic changes to a follow-up PR?

Comment (ngimel) [reply aten/src/ATen/native/cuda/ConvolutionMM2d.cu:205]: sure! Then I only have questions about 3d inputs and 2d weights.

File: aten/src/ATen/native/cuda/ConvolutionMM2d.cu (hunk 1 above)
Comment (ngimel) [reply aten/src/ATen/native/cuda/ConvolutionMM2d.cu:41]: Yes, please do

**Comment by ngimel** [comment]
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30441307).

**Pull request closed without merging**